README.md
# idkit_extension

### Introduction

This package is mainly used to extend common functions of some classes in Flutter development, making it more convenient and simple for developers to develop.

### Functions of currently extended classes

- #### String

  ##### 1. Remove all spaces in the string.

  ```dart
  /// 1. Remove all spaces in the string.
  void removeAllSpaces() {
    const String testStr = ' aa bb  cc dd  ';
    print(testStr.trimAll()); // aabbccdd
  }
  ```

  ##### 2. Convert a string to a number.

  ```dart
  /// 2. Convert a string to a number.
  void toNumber() {
    const String value = '110';
    print(value.toInt()); // 110
    print(value.toDouble()); // 110.0
  }
  ```

  ##### 3. Convert a string to a number.

  ```dart
  /// 3. Convert string to collection.
  void toCollection() {
    const String testStr = 'abcd';
    print(testStr.toList()); // [a,b,c,d]

    const String testStr1 = '1,2,10,21';
    print(testStr1.toList(patternSplit: ',')); // [1,2,10,21]
  }
  ```

  ##### 4. ID card verification.

  ```dart
  /// 4. ID card verification.
  void toID() {
      const String id1 = '53010219200508011X';
      print(id1.isIDCard()); // true

      const String id2 = '53010219200508012X';
      print(id2.isIDCard()); // false
  }
  ```

  ##### 5. Verify that the string is all Chinese characters.

  ```dart
  /// 5. Verify that the string is all Chinese characters.
  void toCC() {
      const String txt1 = '我是who';
      print(txt1.isChineseCharacters()); // false

      const String txt2 = '我是 大鲨鱼，是吗';
      print(txt2.isChineseCharacters()); // true
  }
  ```

  ##### 6. Verify that the string is a qualified phone number.

  ```dart
  /// 6. Verify that the string is a qualified phone number.
  void toPhone() {
      const String phone = '18801210281';
      print(phone.isPhoneNumber()); // true

      const String phone1 = '12801210281';
      print(phone1.isPhoneNumber()); // false
  }
  ```

  ##### 7. Verify whether it is a string composed of 0-9.

  ```dart
  /// 7. Verify whether it is a string composed of 0-9.
  void toVerifyNumber() {
      const String phone = '18801210281';
      print(phone.isAllNumber()); // true

      const String phone1 = '12801210281#';
      print(phone1.isAllNumber()); // false
  }
  ```

  ##### 8. Amount formatted in thousandths.

  ```dart
  /// 8. Amount formatted in thousandths.
  void toAmount() {
      const String amount = '12345';
      print(amount.thousands()); // 12,345

      const String amount1 = '12345.087';
      print(amount1.thousands()); // 12,345.087
  }
  ```

- #### List

  ##### 1. Splicing between collection elements.

  ```dart
  /// 1. Splicing between collection elements.
  void toStringFromList() {
      const List<String> list = <String>['T', 'h', 'i', 's'];
      print(list.splicing()); // This

      const List<String> list1 = <String>['2021', '11', '17'];
      print(list1.splicing('-')); // 2021-11-17
  }
  ```

# - 0.0.3 Update content introduction

#### 1. Add a method to copy the value of the collection object

```dart
void listCopyValue() {
  final List<int> a = <int>[1, 2, 3];
  final List<int> b = a.copy();
  a.add(4);
  print(a); // [1,2,3,4]
  print(b); // [1,2,3]
}
```

#### 2. Copy the value of the key-value collection

```dart
void mapCopy() {
  final Map<String, String> a = <String, String>{'a': '10', 'b': '20'};
  final Map<String, String> b = a;
  final Map<String, String> c = a.copy();
  a['a'] = '00';
  print(a); // {a: 00, b: 20}
  print(b); // {a: 00, b: 20}
  print(c); // {a: 10, b: 20}
}
```

#### 3. A key-value collection gets a new key-value from another key-value collection

```dart
void mapCopyMap() {
  final Map<String, String> a = <String, String>{'a': '10'};
  final Map<String, String> b = <String, String>{
    'a': '00',
    'b': '10',
    'c': '20'
  };
  final Map<String, String> c = a.copyFrom(b);
  print(c); // {a: 10, b: 10, c: 20}
}
```


CHANGELOG.md
## 0.0.1

- Extend the functions of some classes of flutter.
  - Extension of the string class.
    1. Remove all spaces in the string.
    2. The string is directly converted to int or double.
    3. String directly converted to collection sequence.
    4. Can check whether the string is a legal ID number.
    5. It can be verified whether the string is all Chinese characters.
    6. It is possible to check whether the character string is a legal mobile phone number.
    7. It is possible to check whether the character string is all digits.
    8. You can format the string amount.
  - Extension of the list class.
    1. The elements in the collection can be spliced into strings.
- The extension method of each extension class

  - String

    ```dart
    1. trimAll() -> String
    2. toInt() -> int
    3. toDouble() -> double
    4. toList({int? lengthSplit,Pattern? patternSplit}) -> List<String>
    5. isIDCard() -> bool
    6. isChineseCharacters() -> bool
    7. isPhoneNumber() -> bool
    8. isAllNumber() -> bool
    9. thousands() -> String
    ```

  - List
    ```dart
    1. splicing() -> String
    ```

## 0.0.2

- New method extension.

  - String

    1. Convert a number string to a string with specified decimal places.

  - Num
    1. Convert a number to a string with specified decimal places.

- Extension methods of each new extension class.

  - String

    ```dart
    1. keepFractionDigits(int,{bool}) -> String
    ```

  - Num

    ```dart
    1. keepFractionDigits(int,{bool}) -> String
    ```

## 0.0.3

- New method extension.

  - List
    1. Collection object value copy
  - Map
    1. Copy the value of the key-value collection
    2. A key-value collection gets a new key-value from another key-value collection

- Extension methods of each new extension class.
  - List
    ```dart
    1. copy({bool}) -> List
    ```
  - Map
    ```dart
    1. copy() -> Map
    2. copyFrom() -> Map
    ```


test/idkit_extension_test.dart
import 'package:idkit_extension/idkit_extension.dart';

void main() {
  final Map<String, String> a = <String, String>{'a': '10'};
  final Map<String, String> b = <String, String>{
    'a': '00',
    'b': '10',
    'c': '20'
  };
  final Map<String, String> c = a.copyFrom(b);
  print(c); // {a: 10, b: 20}
}


lib/src/list_extension.dart
extension IDKitList<E> on List<E> {
  /// Splicing between collection elements.
  /// [separator] Separator between elements.
  String splicing([String separator = '']) {
    late final StringBuffer buffer = StringBuffer();
    if (isNotEmpty) {
      if (length == 1) {
        buffer.write(first);
      } else {
        buffer.writeAll(this, separator);
      }
    }
    return buffer.toString();
  }

  /// Copy the value of the object.
  /// [growable] Can the generated collection be variable.
  ///
  /// The method is to solve the following problems:
  /// ```dart
  /// List<int> a = <int>[1,2,3];
  /// List<int> b = a;
  /// a.add(4);
  /// print(a); // [1,2,3,4]
  /// print(b); // [1,2,3,4]
  /// ```
  List<E> copy({bool growable = true}) {
    return List<E>.from(this, growable: growable);
  }
}


lib/src/string_extension.dart
import 'package:idkit_extension/src/list_extension.dart';

/// Extend the string class.
extension IDKitString on String {
  /// Remove all spaces in the string.
  String trimAll() => replaceAll(RegExp(r'\s+'), '');

  /// String to numeric value.
  int toInt() => int.tryParse(this) ?? 0;
  double toDouble() => double.tryParse(this) ?? 0.0;

  /// Convert string to collection.
  /// [lengthSplit]  Split string by length.
  /// [patternSplit] Split the string according to the specified pattern.
  /// Note: [lengthSplit] has higher priority than [patternSplit].
  List<String> toList({
    int? lengthSplit,
    Pattern? patternSplit,
  }) {
    late List<String> resultList = <String>[];
    if (lengthSplit != null) {
      final int length = this.length;
      if (lengthSplit == 0 || length < lengthSplit) {
        resultList = <String>[this];
      } else {
        final int remainder = length % lengthSplit;
        final int count = length ~/ lengthSplit;
        resultList = List<String>.generate(count, (int i) {
          final int start = i * lengthSplit;
          return substring(start, start + lengthSplit);
        }).toList();
        if (remainder != 0) {
          resultList.add(substring(length - remainder));
        }
      }
    } else {
      resultList = split(patternSplit ?? '');
    }
    return resultList;
  }

  /// ID card verification.
  ///
  /// Note: Hong Kong, Macau, and Taiwan ID card verification is not currently supported.
  bool isIDCard() {
    late bool result = false;
    if (isNotEmpty) {
      final RegExp regExp = RegExp(r'\d{17}[0-9Xx]$');
      if (regExp.hasMatch(this)) {
        final String forward17 = substring(0, 17);
        final String behind1 = substring(17).toUpperCase();
        final List<String> splitList = forward17.toList();
        // Coefficient for id.
        final List<String> modulusList =
            '7,9,10,5,8,4,2,1,6,3,7,9,10,5,8,4,2'.toList(patternSplit: ',');
        late int sum = 0;
        for (int i = 0; i < 17; i++) {
          final int item1 = int.parse(splitList[i]);
          final int item2 = int.parse(modulusList[i]);
          sum += item1 * item2;
        }
        final int remainder = sum % 11;
        late final Map<String, String> mapping = <String, String>{
          '0': '1',
          '1': '0',
          '2': 'X',
          '3': '9',
          '4': '8',
          '5': '7',
          '6': '6',
          '7': '5',
          '8': '4',
          '9': '3',
          '10': '2'
        };
        result = mapping[remainder.toString()] == behind1;
      }
    }
    return result;
  }

  /// Verify that the string is all Chinese characters.
  bool isChineseCharacters() => RegExp(r'[\u4e00-\u9fa5]+$').hasMatch(this);

  /// Verify that the string is a qualified phone number.
  /// Note: Except for IoT cards and data cards.
  bool isPhoneNumber() => RegExp(
          r'^((13[^4]|14[5|7|9]|15[^4]|166|17[^0149]|18\d|19[^4])\d{8}|134[^9]\d{7})$')
      .hasMatch(this);

  /// Verify whether it is a string composed of 0-9.
  bool isAllNumber() => RegExp(r'\d+$').hasMatch(this);

  /// Amount formatted in thousandths.
  String thousands() {
    late String result = this;
    if (length != 0) {
      late String forward = this;
      late String behind = '';
      if (forward.contains('.')) {
        final List<String> splitList = result.split(r'.');
        forward = splitList.first;
        if (splitList.length >= 2) {
          behind = splitList.last;
        }
      }
      final int forwardLength = forward.length;
      final int remainder = forwardLength % 3;
      late List<String> forwardList = <String>[];
      if (remainder != 0) {
        forward = '0' * (3 - remainder) + forward;
      }
      forwardList = forward.toList(lengthSplit: 3);
      final int firstValue = forwardList.first.toInt();
      forwardList.replaceRange(0, 1, <String>[firstValue.toString()]);
      result = forwardList.splicing(',');
      if (behind.isNotEmpty) {
        result += '.' + behind;
      }
    }
    return result;
  }

  /// Convert a number string to a string with specified decimal places.
  /// [fractionDigits] : Number of decimal places.
  /// [round] : Carry processing, default true carry , false no carry.
  String keepFractionDigits(int fractionDigits, {bool round = true}) {
    late String result = this;
    final bool regular = double.tryParse(this) == null;
    if (regular) {
      // 不合法
      if (fractionDigits != null && fractionDigits != 0) {
        result = '0.' + '0' * fractionDigits;
      } else {
        result = '0';
      }
    } else {
      final bool point = contains(RegExp(r'\.'));
      if (point) {
        final List<String> list = split(RegExp(r'\.'));
        late final String firstPart = list.first;
        late final String lastPart = list.last;
        if (round) {
          // 进位处理
          final double value = double.parse(this);
          result = value.toStringAsFixed(fractionDigits);
        } else {
          // 不进位
          result = firstPart + '.' + lastPart.substring(0, fractionDigits);
        }
      } else {
        // 不含小数
        if (fractionDigits != 0) {
          result += '.' + '0' * fractionDigits;
        }
      }
    }
    return result;
  }
}


example/lib/main.dart
import 'package:idkit_extension/idkit_extension.dart';

void main() {
  removeAllSpaces();
  toNumber();
  toCollection();
  toID();
  toCC();
  toPhone();
  toVerifyNumber();
  toAmount();
  toStringFromList();
  formStringToKeepPoint();
  formNumToKeepPoint();
  listCopyValue();
  mapCopy();
}

/// 1. Remove all spaces in the string.
void removeAllSpaces() {
  const String testStr = ' aa bb  cc dd  ';
  print(testStr.trimAll());
}

/// 2. Convert a string to a number.
void toNumber() {
  const String value = '110';
  print(value.toInt()); // 110
  print(value.toDouble()); // 110.0
}

/// 3. Convert string to collection.
void toCollection() {
  const String testStr = 'abcd';
  print(testStr.toList()); // [a,b,c,d]

  const String testStr1 = '1,2,10,21';
  print(testStr1.toList(patternSplit: ',')); // [1,2,10,21]
}

/// 4. ID card verification.
void toID() {
  const String id1 = '53010219200508011X';
  print(id1.isIDCard()); // true

  const String id2 = '53010219200508012X';
  print(id2.isIDCard()); // false
}

/// 5. Verify that the string is all Chinese characters.
void toCC() {
  const String txt1 = '我是who';
  print(txt1.isChineseCharacters()); // false

  const String txt2 = '我是 大鲨鱼，是吗';
  print(txt2.isChineseCharacters()); // true
}

/// 6. Verify that the string is a qualified phone number.
void toPhone() {
  const String phone = '18801210281';
  print(phone.isPhoneNumber()); // true

  const String phone1 = '12801210281';
  print(phone1.isPhoneNumber()); // false
}

/// 7. Verify whether it is a string composed of 0-9.
void toVerifyNumber() {
  const String phone = '18801210281';
  print(phone.isAllNumber()); // true

  const String phone1 = '12801210281#';
  print(phone1.isAllNumber()); // false
}

/// 8. Amount formatted in thousandths.
void toAmount() {
  const String amount = '12345';
  print(amount.thousands()); // 12,345

  const String amount1 = '12345.087';
  print(amount1.thousands()); // 12,345.087
}

/// 9. Splicing between collection elements.
void toStringFromList() {
  const List<String> list = <String>['T', 'h', 'i', 's'];
  print(list.splicing()); // This

  const List<String> list1 = <String>['2021', '11', '17'];
  print(list1.splicing('-')); // 2021-11-17
}

/// 11. Convert a string of numbers to a number with specified decimal places
void formStringToKeepPoint() {
  final String a = '111.03'.keepFractionDigits(0);
  print(a);

  final String b = '111.03'.keepFractionDigits(1);
  print(b);

  final String c = '111.03'.keepFractionDigits(2);
  print(c);

  final String d = '111.03'.keepFractionDigits(3);
  print(d);

  final String e = '111'.keepFractionDigits(2);
  print(e);

  final String f = '111.034'.keepFractionDigits(2);
  print(f);

  final String f1 = '111.036'.keepFractionDigits(2);
  print(f1);

  final String f2 = '111.036'.keepFractionDigits(2, round: false);
  print(f2);

  final String g = '111.996'.keepFractionDigits(2);
  print(g);

  final String g1 = '111.994'.keepFractionDigits(2);
  print(g1);

  final String h = 'xx23343'.keepFractionDigits(3);
  print(h);
}

/// 12. Convert a string of numbers to a number with specified decimal places
void formNumToKeepPoint() {
  final String a = 111.03.keepFractionDigits(0);
  print(a);

  final String b = 111.03.keepFractionDigits(1);
  print(b);

  final String c = 111.03.keepFractionDigits(2);
  print(c);

  final String d = 111.03.keepFractionDigits(3);
  print(d);

  final String e = 111.keepFractionDigits(2);
  print(e);

  final String f = 111.034.keepFractionDigits(2);
  print(f);

  final String f1 = 111.036.keepFractionDigits(2);
  print(f1);

  final String f2 = 111.036.keepFractionDigits(2, round: false);
  print(f2);

  final String g = 111.996.keepFractionDigits(2);
  print(g);

  final String g1 = 111.994.keepFractionDigits(2);
  print(g1);
}

/// 12. Collection object value copy
void listCopyValue() {
  final List<int> a = <int>[1, 2, 3];
  final List<int> b = a.copy();
  a.add(4);
  print(a); // [1,2,3,4]
  print(b); // [1,2,3]
}

/// 13. Copy the value of the key-value collection
void mapCopy() {
  final Map<String, String> a = <String, String>{'a': '10', 'b': '20'};
  final Map<String, String> b = a;
  final Map<String, String> c = a.copy();
  a['a'] = '00';
  print(a); // {a: 00, b: 20}
  print(b); // {a: 00, b: 20}
  print(c); // {a: 10, b: 20}
}

/// 14. A key-value collection gets a new key-value from another key-value collection
void mapCopyMap() {
  final Map<String, String> a = <String, String>{'a': '10'};
  final Map<String, String> b = <String, String>{
    'a': '00',
    'b': '10',
    'c': '20'
  };
  final Map<String, String> c = a.copyFrom(b);
  print(c); // {a: 10, b: 10, c: 20}
}
